Cert manager changes (#18023)
cockroachdb/docs: src/current/v23.1/secure-cockroachdb-kubernetes.md

- [Secure the webhooks](#secure-the-webhooks) (Operator)

{{site.data.alerts.callout_danger}}

> If you are running a secure Helm deployment on Kubernetes 1.22 and later, you must migrate away from using the Kubernetes CA for cluster authentication. The recommended approach is to use `cert-manager` for certificate management. For details, see [Deploy cert-manager for mTLS](#deploy-cert-manager-for-mtls).

Reviewer: ```suggestion
If you are running a secure Helm deployment on Kubernetes 1.22 and later, you must migrate away from using the Kubernetes CA for cluster authentication. The recommended approach is to use `cert-manager` for certificate management. For details, refer to [Deploy cert-manager for mTLS].
```

    The certificates will be rotated during the specified expiry windows.

> ## Deploy cert-manager for mTLS
> 
> The recommended approach for cluster authentication is to use cert-manger to sign certificates. cert-manager adds certificates and certificate issuers as resource types in Kubernetes clusters, and simplifies the process of obtaining, renewing and using those certificates.
> 
> 1. Make sure you have the latest version of cert-manger installed you will find it [here](https://cert-manager.io/docs/installation/).
> 
> 2. The first thing you will need to do is to create an `Issuer` these are the resources that represent certificate authorities and are able to sign certificates. Using cert-manger create an `Issuer` for signing self-signed CA certificate.

Reviewer: ```suggestion
## Deploy `cert-manager` for mTLS

Cockroach Labs recommends using `cert-manager` to sign certificates for cluster authentication. `cert-manager` manages certificates and certificate issuers as resource types in Kubernetes clusters, to simplify the process of obtaining, renewing and using those certificates.

1. Install a [supported version of `cert-manger`](https://cert-manager.io/docs/releases/). For a new cluster, Cockroach Labs recommends using the latest supported version. Refer to installed you will find it [`cert-manager` Installation](https://cert-manager.io/docs/installation/) in the `cert-manager` project's documentation.

2. Create a file named `issuer.yaml` that configures an `Issuer`, which represents a certificate authority that can sign certificates. This example creates an issuer that can sign self-signed CA certificates. To customize your issuer, refer to [Issuer Configuration](https://cert-manager.io/docs/configuration/) in the `cert-manager` project's documentation.
```

    {% include_cached copy-clipboard.html %}
    ~~~yaml
    apiVersion: cert-manager.io/v1
    kind: Issuer
    metadata:
    name: cockroachdb
    spec:
    selfSigned: {}

> 3. Apply this to your Kubernetes cluster using `kubectl`

Reviewer: ```suggestion
3. Use `kubectl apply` to create the issuer from the YAML file:

    {% include_cached copy-clipboard.html %}
    ~~~shell
    kubectl apply -f issuer.yaml
    ~~~

```

> 4. To enable `cert-manager` you will need to have the following settings in the values.yaml file.

Reviewer: ```suggestion
4. Create a file named `cert-manager.yaml` that enables signing self-signed certificates and configures the `cert-manager` Kubernetes resource.
```

> 4. To enable `cert-manager` you will need to have the following settings in the values.yaml file.

Reviewer: ```suggestion
4. Enable and configure `cert-manager` in the Helm chart's `values.yaml` file. The following options are required. For more options, refer to [`cert-manager`](https://github.com/cockroachdb/helm-charts/tree/master/cockroachdb#cert-manager) in the CockroachDB Helm chart documentation.
```

    {% include_cached copy-clipboard.html %}
    ~~~yaml
    tls.certs.selfSigner.enabled: false
    tls.certs.certManager: true
    tls.certs.certManagerIssuer.kind: Issuer
    tls.certs.certManagerIssuer.name: cockroachdb
    ~~~

> - Setting `tls.certs.selfSigner.enabled` to `false` will disable the self signing certificates for cockroachdb.
> - By enabling the`tls.certs.certManager:` setting to true this will enable cert-manager for certificate signing.
> - Specify the ‘Issuer` kind here either `Issuer` or `ClusterIssuer` under the setting `certManagerIssuer.kind`.
> - `certManagerIssuer.name` Provide the Issuer you have created in previous step. In this example it was `cockroachdb`.

Reviewer: ```suggestion
    - To disable signing self-signed certificates, set `tls.certs.selfSigner.enabled` to `false`.
    - Set `tls.certs.certManagerIssuer.kind` to either `Issuer` or `ClusterIssuer`. To get started, `Issuer` is recommended. `ClusterIssuer` is cluster-scoped; when referencing a secret via the `secretName` field, only secrets in the `cluster-resource` namespace (`cert-manager` by default) are searched. To learn more, refer to [Cluster Resource Namespace](https://cert-manager.io/v1.6-docs/faq/cluster-resource/) in the `cert-manager` project's documentation.
    - Set `certManagerIssuer.name` to the name of the issuer you created in the previous step. 
```

> 5. Once the values.yaml file has been updated you can then deploy CockroachDB via the Helm Chart with the following command.

Reviewer: ```suggestion
5. Apply the updated Helm chart:
```

    {% include_cached copy-clipboard.html %}
    ~~~shell

>     helm install my-release --values {custom-values}.yaml cockroachdb/cockroachdb

Reviewer: ```suggestion
    helm install my-release --values values.yaml cockroachdb/cockroachdb
```

    ~~~

> 

Reviewer: ```suggestion

Replace `values.yaml` with the name of your Helm chart's values file.
```

> ## Migration to self-signer

Reviewer: Should this section be removed?

Reviewer 2: I think that it should be removed and cert-manager adopted moving forward.
